src/rest_vol_datatype.c

```diff
@@ -2609,3 +2609,174 @@ RV_tconv_init(hid_t src_type_id, size_t *src_type_size, hid_t dst_type_id, size_
+    char       *src_member_name = NULL, *dst_member_name = NULL;
+    size_t      src_member_offset = 0, dst_member_offset = 0;
+    hid_t       src_member_type = H5I_INVALID_HID, dst_member_type = H5I_INVALID_HID;
+    int         dst_nmembs = 0, src_nmembs = 0;
+    hbool_t     member_match = FALSE;
+
+    if (H5T_NO_CLASS == (src_type_class = H5Tget_class(src_type_id)))
+        FUNC_GOTO_ERROR(H5E_DATATYPE, H5E_BADVALUE, FAIL, "source datatype is invalid");
+
+    if (H5T_NO_CLASS == (dst_type_class = H5Tget_class(dst_type_id)))
+        FUNC_GOTO_ERROR(H5E_DATATYPE, H5E_BADVALUE, FAIL, "destination datatype is invalid");
+
+    if ((dst_type_class != H5T_COMPOUND) || (src_type_class != H5T_COMPOUND)) {
+        *subset = H5T_SUBSET_FALSE;
+        FUNC_GOTO_DONE(SUCCEED);
+    }
+
+    if ((dst_nmembs = H5Tget_nmembers(dst_type_id)) < 0)
+        FUNC_GOTO_ERROR(H5E_DATATYPE, H5E_CANTGET, FAIL, "can't get nmembers of destination datatype");
+
+    if ((src_nmembs = H5Tget_nmembers(src_type_id)) < 0)
+        FUNC_GOTO_ERROR(H5E_DATATYPE, H5E_CANTGET, FAIL, "can't get nmembers of source datatype");
+
+    /* Determine subset status by comparing members */
+    if (src_nmembs < dst_nmembs) {
+        for (unsigned src_idx = 0; src_idx < src_nmembs; src_idx++) {
+            member_match = FALSE;
+
+            if ((src_member_type = H5Tget_member_type(src_type_id, src_idx)) == H5I_INVALID_HID)
+                FUNC_GOTO_ERROR(H5E_DATATYPE, H5E_CANTGET, FAIL, "can't get memory datatype member type");
+
+            if ((src_member_name = H5Tget_member_name(src_type_id, src_idx)) == NULL)
+                FUNC_DONE_ERROR(H5E_DATATYPE, H5E_CANTGET, FAIL, "can't get memory datatype member name");
+
+            if ((src_member_offset = H5Tget_member_offset(src_type_id, src_idx)) < 0)
+                FUNC_DONE_ERROR(H5E_DATATYPE, H5E_CANTGET, FAIL, "can't get memory datatype member offset");
+
+            /* Search for match in dst compound */
+            for (unsigned dst_idx = 0; dst_idx < dst_nmembs; dst_idx++) {
+                if ((dst_member_type = H5Tget_member_type(dst_type_id, dst_idx)) == H5I_INVALID_HID)
+                    FUNC_GOTO_ERROR(H5E_DATATYPE, H5E_CANTGET, FAIL,
+                                    "can't get destination datatype member type");
+
+                if ((dst_member_name = H5Tget_member_name(dst_type_id, dst_idx)) == NULL)
+                    FUNC_DONE_ERROR(H5E_DATATYPE, H5E_CANTGET, FAIL,
+                                    "can't get destination datatype member name");
+
+                if ((dst_member_offset = H5Tget_member_offset(dst_type_id, dst_idx)) < 0)
+                    FUNC_DONE_ERROR(H5E_DATATYPE, H5E_CANTGET, FAIL,
+                                    "can't get destination datatype member offset");
+
+                /* Compare member names and offsets */
+                if (!strcmp(src_member_name, dst_member_name) && (src_member_offset == dst_member_offset)) {
```

Comment:
Checking member names and offsets isn't necessarily enough since the members could be different datatypes; we also need to check if they're equal with `H5Tequal`. Unfortunate that this logic had to be re-implemented from what the library does, but maybe we should create an issue about exposing this subset detection functionality publicly for VOL connectors to use.